File: lib/ch2/2.4.js
// You have two numbers represented by a linked list, where each node contains a sin- gle digit The digits are stored in reverse order, such that the 1’s digit is at the head of the list Write a function that adds the two numbers and returns the sum as a linked list
// EXAMPLE
// Input: (3 -> 1 -> 5) + (5 -> 9 -> 2)
// Output: 8 -> 0 -> 8

const Node = require("./helpers/Node");

const add = (lst1, lst2) => {
  const lst3 = new Node(0);
  let n1 = lst1;
  let n2 = lst2;
  let n3 = lst3;
  let rem = 0;

  while (n1 || n2) {
    const v1 = n1 ? n1.value : 0;
    const v2 = n2 ? n2.value : 0;
    n3.value = (v1 + v2 + rem) % 10;
    rem = Math.floor((v1 + v2 + rem) / 10);
    if (!n1.next && !n2.next) break;
    n3.next = new Node(0);
    n1 = n1.next || false;
    n2 = n2.next || false;
    n3 = n3.next;
  }

  if (rem !== 0) n3.next = new Node(rem);

  return lst3;
};

module.exports = add;

File: java/ch2/2.4.java
public Node sumLinkedList(Node n1, Node n2) {
    Node sentinel = new Node(0);
    Node n3 = sentinel;
    int carry = 0;
    while (n1 != null || n2 != null) {
        int d1 = (n1 == null ? 0 : n1.data);
        int d2 = (n2 == null ? 0 : n2.data);
        int sum = d1 + d2 + carry;
        n3.next = new Node(sum % 10);
        n3 = n3.next;
        carry = (sum >= 10 ? 1 : 0);
        if (n1 != null) n1 = n1.next;
        if (n2 != null) n2 = n2.next;
    }
    if (carry != 0) n3.next = new Node(carry);
    return sentinel.next;
}
